gr-uhd/lib/rfnoc_fir_filter_impl.cc

```diff
@@ -0,0 +1,81 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2022 Ettus Research, A National Instruments Brand.
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "rfnoc_fir_filter_impl.h"
+#include <gnuradio/io_signature.h>
+#include <uhd/version.hpp>
+#include <volk/volk.h>
+
+#if UHD_VERSION >= 4020000
+#define HAVE_FIR_FILTER_MULTICHAN
+#else
+#define CHECK_CHAN_VALID(chanvar)                                             \
+    if (chanvar > 0) {                                                        \
+        d_logger->warn("Invalid channel index: {}. This version of UHD only " \
+                       "supports one channel.",                               \
+                       chanvar);                                              \
+    }
+#endif
+
+namespace gr {
+namespace uhd {
+
+rfnoc_fir_filter::sptr rfnoc_fir_filter::make(rfnoc_graph::sptr graph,
+                                              const ::uhd::device_addr_t& block_args,
+                                              const int device_select,
+                                              const int instance)
+{
+    return gnuradio::make_block_sptr<rfnoc_fir_filter_impl>(
+        rfnoc_block::make_block_ref(graph, block_args, "FIR", device_select, instance));
+}
+
+
+rfnoc_fir_filter_impl::rfnoc_fir_filter_impl(::uhd::rfnoc::noc_block_base::sptr block_ref)
+    : rfnoc_block(block_ref),
+      d_fir_filter_ref(get_block_ref<::uhd::rfnoc::fir_filter_block_control>())
+{
+}
+
+rfnoc_fir_filter_impl::~rfnoc_fir_filter_impl() {}
+
+/******************************************************************************
+ * rfnoc_fir_filter API
+ *****************************************************************************/
+void rfnoc_fir_filter_impl::set_coefficients(const std::vector<float>& coeffs,
+                                             const size_t chan = 0)
+{
+    std::vector<int16_t> coeffs_int16(coeffs.size());
+
+    const float* in = &coeffs.front();
+    int16_t* out = &coeffs_int16.front();
+
+    volk_32f_s32f_convert_16i(out, in, 32768.0, coeffs.size());
+    set_coefficients(coeffs_int16, chan);
+}
+
+void rfnoc_fir_filter_impl::set_coefficients(const std::vector<int16_t>& coeffs,
+                                             const size_t chan = 0)
+{
+    d_fir_filter_ref->set_coefficients(coeffs, chan);
```

Comment:
This leads to build failures on machines with UHD 4 < 4.2, as far as I can tell.

The missing second parameter `chan` was added in `UHD-4.2`. Which is a bit sad, as that kind of says this needs to be `#if`'ed to match the UHD system that we're building against, but since we don't want the UHD version to bleed into the GNU Radio API, we need to lug around the (then dead) `chan` parameter in `void rfnoc_fir_filter_impl::set_coefficients(const std::vector<int16_t>& coeffs, const size_t chan = 0)` and log a warning.


```suggestion
#if UHD_VERSION >= 4020000
    d_fir_filter_ref->set_coefficients(coeffs, chan);
#else
    d_logger->warn("Trying to set coefficients for channel {:d}. Versions prior to UHD 4.2 do not support per-channel setting.", chan);
    d_fir_filter_ref->set_coefficients(coeffs);
```

Reply:
Well look at me.. I forgot to port half of the code from gr-ettus. See the stuff at the top of this file, that's what's supposed to go here.

Reply:
:) so you kinda did port it :)